File: 17.letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        strs = []
        if not digits:
            return strs
        init_str = ""
        lut = {
            '2': "abc",
            '3': "def",
            '4': "ghi",
            '5': "jkl",
            '6': "mno",
            '7': "pqrs",
            '8': "tuv",
            '9': "wxyz",
        }
        for c in digits:
            init_str +=  lut[c][0]
        strs.append(init_str)
        for i in range(len(digits)):
            for str in strs[:len(strs)]:
                replace = lut[digits[i]]
                for c in replace[1:]:
                    strs.append(str[:i] + c + str[i + 1:])
        return strs

File: 17.letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> strs;
        if (digits.empty()) {
            return strs;
        }
        string init_str = "";
        map<char, string> lut{
            {'2', "abc"},
            {'3', "def"},
            {'4', "ghi"},
            {'5', "jkl"},
            {'6', "mno"},
            {'7', "pqrs"},
            {'8', "tuv"},
            {'9', "wxyz"},
        };
        for (int i = 0; i < digits.length(); i++) {
            init_str +=  lut[digits[i]][0];
        }
        strs.push_back(init_str);
        for (int i = 0; i < digits.length(); i++) {
            int n = strs.size();
            for (int j = 0; j < n; j++) {
                string str = strs[j];
                string replace = lut[digits[i]];
                for (int k = 1; k < replace.length(); k++) {
                    str[i] = replace[k];
                    strs.push_back(str);
                }
            }
        }
        return strs;
    }
};